design/external-compaction.md

```diff
@@ -0,0 +1,249 @@
+ */
+CompactionJob getCompactionJob(String queueName, CompactorId)
+
+
+/**
+ * Called by compactor process to indicate work is done. The implementation will call  finishedJob on the tablet servers.
+ **/
+void finishedJob(CompactionJob, CompactionStats)
+
+
+/**
+  * Called by compactor process to update the coordinator with the state of the running compaction
+ **/
+void updateCompactionStatus(CompactionJob, State, message, timestamp)
+```
+
+### Compactor RPC API
+
+```
+/**
+  * Called by CompactionCoordinator to cancel the compaction work
+**/
+void cancel(CompactionJob)
+```
+
+### Tablet Server Compaction RPC API
+
+```
+/**
+  Summary information about a tablet servers external compaction queue.
+ */
+Class TabletServerCompactionQueueSummary {
+  // the name of an external queue for which information is summarized
+  Sting queueName;
+   // the priority for which this information is summarized
+   Long priority;
+   // the number of tablets queued on a tserver for a given queue and priority
+   Int queuedCount;
+}
+
+
+/**
+ * Called by the coordinator to get summary information about all of a tablets servers external compaction queues.
+ */
+List<TabletServerCompactionQueueSummary> getTabletServerQueueInfo()
+
+
+/**
+ * Called to atomically reserve compaction work for a tablet on a given queue with a priority level >= than what was passed.  Will return null if there is no work meeting criteria.  Tablets will treat these files reserved as part of a running compaction for planning future compactions. Will use CompactorId to check if process running compactions is alive.  Information about the reservation will be persisted by the tablet in the metadata table so that if the tablet loads elsewhere it will not lose track of the external compaction.
+ */
+CompactionJob reserveJob(String queueName, long priority, CompactorId)
+
+
+/**
+ * Atomically updates a tablets files after a compaction finishes. Called by CompactionCoordinator when Compactor finishes compaction
+ */
+void finishedJob(CompactionJob)
+```
+
+## Queue Specification
+
+The CompactionCoordinator is responsible for managing the work queue. For each external compaction queue, the tablet server will maintain an internal in memory priority queue of the tablets loaded on it that have work to do on that external queue. The coordinator polls all tservers to get summary information about their external compaction queues. The coordinator then combines the summary information from all tservers and uses it to determine which tablet server to contact next to get work.  The coordinator does not maintain per tablet information, it only maintains enough information to allow it know which tablet server to contact for a given queue.  The tablet server will then know what specific tablet in that queue needs to compact.  
+
```

Comment:
Would it be better to have the tservers push the summary information when it changes? This way we don't want to burden tservers by polling while it is busy.

Reply:
> Would it be better to have the tservers push the summary information when it changes?

I am uncertain which is better, but I am leaning a bit towards pushing.  One advantage to tservers pushing is that tservers w/ nothing in their external queues would not push anything.  With polling, it would keep polling tservers that have no tablets in their external queues.

Was thinking polling is better in the case when the coordinator starts up.  It can do something like the following.

 * poll all tservers
 * start accepting request for work from compactors

With push, I suppose it could do something like the following

 * wait X minutes for tservers to report external compaction summary info
 * start accepting request for work from compactors

So need to figure out the coordinator startup behavior w/ push.

Reply:
What about only polling at startup so compactors can get started right away? I suppose a hybrid approach would then require 2 different RPC endpoints though. Maybe that's not such a big deal?

Reply:
The hybrid approach may be best, I was hoping to think of something simpler.

Reply:
If the tserver push frequency is known, like tservers push at least every 30 secs.  Then the coordinator could wait 2x that I think and it would be close to optimal.  If the coordinator misses a few updates from tservers before handing out work, then it may start some lower priority compactions running before higher prio ones.  That is probably ok.